### DISCUSSION TITLE
Exposing internal properties of AnimationNodeStateMachinePlayback

### DISCUSSION THREAD

**Discussion by Shnazzy** [godotengine/godot-proposals#5511] [Engine Core]
Currently there's no known way to access the internal properties of this class.  This normally isn't a problem for a majority of use cases but it presents a hurdle if you want to store the state of the machine so you can restore it at some later point.  

A few examples of use cases:

* Games with rewind mechanics such as Braid or Prince of Persia: In this case you would store the last couple seconds of the game's state in a buffer and when rewinding, traverse the buffer in the opposite direction.
* An instant replay feature in a sports game so you can reliably reload to a highlighted moment exactly the way it was when it was recorded.
* In my case, I'm making a fighting game where the movement heavily relies on the state machine node to handle the network of states and transitions between each.  I intend to support rollback netcode, which has a phase in which the state of the match is unwound a small amount in order to recalculate something about the "true" flow of events given new information.

My current solution is to just use a custom build that exposes the properties I need from the engine for my own purposes, however I still see value in finding a general way to achieve this since there are other use cases besides my own.  Still, these properties are decently technical, and are likely encapsulated inside the class for good reason.  Simply exposing them might not be the best solution.

My only other idea would be to provide access to this information through a module.  I haven't tested this yet to know if it would work.

**Comment by Calinou** [comment]
Related to https://github.com/godotengine/godot-proposals/issues/4449 (as particle seeking is required for accurate replay playback of particle simulations).

**Comment by TokageItLab** [comment]
I am also interested in the rollback netcode and would like to be able to support it.

However, I am concerned about exposing all properties. Especially the setters is dangerous since it may cause users to break StateMachine. So I think those should be provided as methods such as doing rewind, seek and etc.

**Reply by Shnazzy** [reply]
What about a "PlayBackSnapshot" class or struct that can read or write to the playback object? It would be a friend class to have access to the private properties. Then add a note in the documentation that it's an advanced tool and should be handled carefully.
